File: frontend/src/utils/businessDays.ts
import HolidayJp from '@holiday-jp/holiday_jp'
import { format, parseISO, addDays, isWeekend } from 'date-fns'

/**
 * Check if a given date is a business day (not weekend or Japanese holiday)
 * @param date - Date string in YYYY-MM-DD format or Date object
 * @returns true if the date is a business day
 */
export const isBusinessDay = (date: string | Date): boolean => {
  const dateObj = typeof date === 'string' ? parseISO(date) : date
  
  // Check if it's weekend (Saturday or Sunday)
  if (isWeekend(dateObj)) {
    return false
  }
  
  // Check if it's a Japanese holiday
  const dateString = format(dateObj, 'yyyy-MM-dd')
  const holiday = HolidayJp.isHoliday(new Date(dateString))
  
  return !holiday
}

/**
 * Get all business days between two dates (inclusive)
 * @param startDate - Start date in YYYY-MM-DD format
 * @param endDate - End date in YYYY-MM-DD format
 * @returns Array of Date objects representing business days
 */
export const getBusinessDaysBetween = (startDate: string, endDate: string): Date[] => {
  const businessDays: Date[] = []
  const start = parseISO(startDate)
  const end = parseISO(endDate)
  
  let currentDate = start
  while (currentDate <= end) {
    if (isBusinessDay(currentDate)) {
      businessDays.push(new Date(currentDate))
    }
    currentDate = addDays(currentDate, 1)
  }
  
  return businessDays
}

/**
 * Calculate the ideal line for burndown/burnup charts considering only business days
 * @param totalPoints - Total points to be consumed over the period
 * @param startDate - Start date of the period in YYYY-MM-DD format
 * @param endDate - End date of the period in YYYY-MM-DD format
 * @param chartDates - Array of date strings for which to calculate ideal points
 * @returns Array of ideal point values corresponding to chartDates
 */
export const calculateBusinessDayIdealLine = (
  totalPoints: number,
  startDate: string,
  endDate: string,
  chartDates: string[]
): number[] => {
  // Get all business days in the period
  const businessDays = getBusinessDaysBetween(startDate, endDate)
  const totalBusinessDays = businessDays.length
  
  // If no business days, return array of totalPoints (no progress)
  if (totalBusinessDays === 0) {
    return chartDates.map(() => totalPoints)
  }
  
  // Calculate points consumed per business day
  const pointsPerBusinessDay = totalPoints / totalBusinessDays
  
  // Calculate ideal points for each chart date
  return chartDates.map(dateString => {
    // Count business days from start up to (and including) current date
    const businessDaysCompleted = getBusinessDaysBetween(startDate, dateString).length
    
    // For burndown: remaining points = total - consumed
    // For burnup: planned progress = consumed points
    const consumedPoints = businessDaysCompleted * pointsPerBusinessDay
    
    // Return remaining points for burndown chart
    return Math.max(0, totalPoints - consumedPoints)
  })
}

/**
 * Calculate the ideal line for burnup charts considering only business days
 * @param totalPoints - Total points to be completed over the period
 * @param startDate - Start date of the period in YYYY-MM-DD format
 * @param endDate - End date of the period in YYYY-MM-DD format
 * @param chartDates - Array of date strings for which to calculate ideal points
 * @returns Array of ideal completed point values corresponding to chartDates
 */
export const calculateBusinessDayIdealLineForBurnUp = (
  totalPoints: number,
  startDate: string,
  endDate: string,
  chartDates: string[]
): number[] => {
  // Get all business days in the period
  const businessDays = getBusinessDaysBetween(startDate, endDate)
  const totalBusinessDays = businessDays.length
  
  // If no business days, return array of zeros (no progress)
  if (totalBusinessDays === 0) {
    return chartDates.map(() => 0)
  }
  
  // Calculate points completed per business day
  const pointsPerBusinessDay = totalPoints / totalBusinessDays
  
  // Calculate ideal completed points for each chart date
  return chartDates.map(dateString => {
    // Count business days from start up to (and including) current date
    const businessDaysCompleted = getBusinessDaysBetween(startDate, dateString).length
    
    // For burnup: ideal completed points = accumulated progress
    const completedPoints = businessDaysCompleted * pointsPerBusinessDay
    
    return Math.min(totalPoints, completedPoints)
  })
}

File: backend/app/utils/business_days.py
"""営業日計算ユーティリティ"""

from datetime import date, timedelta
from typing import List
import holidays

class BusinessDayCalculator:
    """営業日計算クラス"""
    
    def __init__(self):
        # 日本の祝日カレンダーを初期化
        self.jp_holidays = holidays.Japan()
    
    def is_business_day(self, target_date: date) -> bool:
        """指定日が営業日かどうかを判定
        
        Args:
            target_date: 判定対象の日付
            
        Returns:
            営業日の場合True、土日祝日の場合False
        """
        # 土日の場合はFalse
        if target_date.weekday() >= 5:  # 5:土曜, 6:日曜
            return False
        
        # 日本の祝日の場合はFalse
        if target_date in self.jp_holidays:
            return False
        
        return True
    
    def get_business_days_between(self, start_date: date, end_date: date) -> List[date]:
        """期間内の営業日リストを取得
        
        Args:
            start_date: 開始日
            end_date: 終了日（含む）
            
        Returns:
            営業日のリスト
        """
        business_days = []
        current_date = start_date
        
        while current_date <= end_date:
            if self.is_business_day(current_date):
                business_days.append(current_date)
            current_date += timedelta(days=1)
        
        return business_days
    
    def count_business_days(self, start_date: date, end_date: date) -> int:
        """期間内の営業日数を計算
        
        Args:
            start_date: 開始日
            end_date: 終了日（含む）
            
        Returns:
            営業日数
        """
        return len(self.get_business_days_between(start_date, end_date))
    
    def calculate_business_day_progress(
        self, 
        start_date: date, 
        end_date: date, 
        current_date: date
    ) -> float:
        """営業日ベースでの進捗率を計算
        
        Args:
            start_date: 期間開始日
            end_date: 期間終了日
            current_date: 現在日
            
        Returns:
            進捗率（0.0〜1.0）
        """
        if current_date <= start_date:
            return 0.0
        if current_date >= end_date:
            return 1.0
        
        total_business_days = self.count_business_days(start_date, end_date)
        if total_business_days == 0:
            return 1.0
        
        elapsed_business_days = self.count_business_days(start_date, current_date)
        return min(1.0, elapsed_business_days / total_business_days)